File: Day #920/solution.js
// Bloom filter: fixed bit array (1000 bits) + k=3 hashes via double hashing.
// add/check are O(k); space O(m) bits. check has false positives, no false negatives.
'use strict';

class BloomFilter {
  constructor() {
    this.SIZE = 1000;
    this.K = 3;
    this.bits = new Array(this.SIZE).fill(false);
  }

  _hash(s, seed) {
    let h = seed >>> 0;
    for (let i = 0; i < s.length; i++) {
      h = Math.imul(h ^ s.charCodeAt(i), 16777619) >>> 0;
    }
    return h >>> 0;
  }

  _baseHashes(v) {
    return [this._hash(v, 2166136261), this._hash(v, 0x811c9dc5 ^ 0x55)];
  }

  add(v) {
    const [h1, h2] = this._baseHashes(v);
    for (let i = 0; i < this.K; i++) this.bits[(h1 + i * h2) % this.SIZE] = true;
  }

  check(v) {
    const [h1, h2] = this._baseHashes(v);
    for (let i = 0; i < this.K; i++) if (!this.bits[(h1 + i * h2) % this.SIZE]) return false;
    return true;
  }
}

function main() {
  const bf = new BloomFilter();
  const added = ['apple', 'banana', 'cherry'];
  added.forEach((s) => bf.add(s));

  console.log('Added values (expect all true):');
  added.forEach((s) => console.log(`  check(${s}) = ${bf.check(s)}`));

  console.log('Non-added values (expect mostly false):');
  ['date', 'elderberry', 'fig', 'grape'].forEach((s) =>
    console.log(`  check(${s}) = ${bf.check(s)}`)
  );
}

main();

File: Day #920/solution.py
# Day 920: Bloom filter: fixed bit array (1000 bits) + k=3 hashes via double hashing.
# add/check are O(k); space O(m) bits. check has false positives, no false negatives.
import hashlib


class BloomFilter:
    SIZE = 1000
    K = 3

    def __init__(self):
        self.bits = [False] * self.SIZE

    def _base_hashes(self, v):
        h1 = int(hashlib.md5(v.encode()).hexdigest(), 16)
        h2 = int(hashlib.sha1(v.encode()).hexdigest(), 16)
        return h1, h2

    def add(self, v):
        h1, h2 = self._base_hashes(v)
        for i in range(self.K):
            self.bits[(h1 + i * h2) % self.SIZE] = True

    def check(self, v):
        h1, h2 = self._base_hashes(v)
        for i in range(self.K):
            if not self.bits[(h1 + i * h2) % self.SIZE]:
                return False
        return True


def main():
    bf = BloomFilter()
    added = ["apple", "banana", "cherry"]
    for s in added:
        bf.add(s)

    print("Added values (expect all true):")
    for s in added:
        print(f"  check({s}) = {bf.check(s)}")

    print("Non-added values (expect mostly false):")
    for s in ["date", "elderberry", "fig", "grape"]:
        print(f"  check({s}) = {bf.check(s)}")


if __name__ == "__main__":
    main()
